File: scripts/compress_assets.py
import os
from PIL import Image

def compress_images(directory, quality=85):
    """
    Compresses all PNG images in the given directory and its subdirectories.
    
    Args:
        directory (str): Path to the directory containing images.
        quality (int): Compression quality (1-100). Default is 85.
    """
    total_savings = 0
    
    for root, dirs, files in os.walk(directory):
        for file in files:
            if file.lower().endswith('.png'):
                file_path = os.path.join(root, file)
                try:
                    # Get original size
                    original_size = os.path.getsize(file_path)
                    
                    with Image.open(file_path) as img:
                        # Convert to RGB (if RGBA) to allow JPEG conversion? No, we need transparency for game assets.
                        # For PNG, we can use optimize=True and reduce colors if needed, 
                        # but simply saving with optimize=True often helps.
                        # Quantizing to 256 colors (P palette) often saves the most space for sprites 
                        # while maintaining transparency.
                        
                        # Check if already P mode (paletted)
                        if img.mode != 'P':
                            # Quantize to reduce colors and size significantly
                            img = img.quantize(colors=256, method=2) # method 2 = FASTOCTREE
                        
                        img.save(file_path, "PNG", optimize=True)
                    
                    # Get new size
                    new_size = os.path.getsize(file_path)
                    savings = original_size - new_size
                    total_savings += savings
                    
                    if savings > 0:
                        print(f"Compressed {file}: {original_size/1024:.1f}KB -> {new_size/1024:.1f}KB (Saved {savings/1024:.1f}KB)")
                    else:
                        print(f"Skipped {file}: No savings")
                        
                except Exception as e:
                    print(f"Error compressing {file}: {e}")

    print(f"\nTotal space saved: {total_savings / (1024*1024):.2f} MB")

if __name__ == "__main__":
    assets_dir = os.path.join(os.getcwd(), 'frontend/public/assets')
    print(f"Compressing images in {assets_dir}...")
    compress_images(assets_dir)

File: scripts/compress_assets.js
import sharp from 'sharp';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Target directory: frontend/public/assets
const ASSETS_DIR = path.resolve(__dirname, '../frontend/public/assets');

async function compressImage(filePath) {
  try {
    const originalBuffer = await fs.promises.readFile(filePath);
    const originalSize = originalBuffer.length;

    // Compress
    const data = await sharp(originalBuffer)
      .png({
        quality: 80,
        compressionLevel: 9,
        palette: true, // Use palette-based compression (like distinct colors) if possible, great for pixel art/sprites
        effort: 10     // Max effort
      })
      .toBuffer();

    const newSize = data.length;
    const savings = originalSize - newSize;

    if (savings > 0) {
      await fs.promises.writeFile(filePath, data);
      console.log(`✅ Compressed ${path.basename(filePath)}: ${(originalSize / 1024).toFixed(1)}KB -> ${(newSize / 1024).toFixed(1)}KB (Saved ${(savings / 1024).toFixed(1)}KB)`);
      return savings;
    } else {
      console.log(`⏩ Skipped ${path.basename(filePath)} (No savings)`);
      return 0;
    }
  } catch (error) {
    console.error(`❌ Error compressing ${path.basename(filePath)}:`, error);
    return 0;
  }
}

async function walkDir(dir) {
  let files = await fs.promises.readdir(dir);
  let totalSavings = 0;

  for (const file of files) {
    const filePath = path.join(dir, file);
    const stat = await fs.promises.stat(filePath);

    if (stat.isDirectory()) {
      totalSavings += await walkDir(filePath);
    } else if (file.toLowerCase().endsWith('.png')) {
      totalSavings += await compressImage(filePath);
    }
  }
  return totalSavings;
}

(async () => {
  console.log(`🚀 Starting compression in ${ASSETS_DIR}...`);
  if (!fs.existsSync(ASSETS_DIR)) {
    console.error(`❌ Directory not found: ${ASSETS_DIR}`);
    process.exit(1);
  }

  const totalSavings = await walkDir(ASSETS_DIR);
  console.log(`\n🎉 Compression Complete! Total space saved: ${(totalSavings / (1024 * 1024)).toFixed(2)} MB`);
})();
